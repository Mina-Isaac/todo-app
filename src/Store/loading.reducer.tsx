import { ActionType, getType } from "typesafe-actions";
import * as postActions from "./Posts/posts.actions";
import { loadTodosAsync } from "./Todos/todos.actions";
export enum loadingFlag {
  unchanged = "unchanged",
  requested = "requested",
  succeded = "succeded",
  failed = "failed",
}
type Action = ActionType<typeof loadTodosAsync | typeof postActions>;
const initialState = {
  todoData: loadingFlag.unchanged,
  postData: loadingFlag.unchanged,
  postAdd: loadingFlag.unchanged,
};

export default function (
  state = initialState,
  action: Action
): typeof initialState {
  switch (action.type) {
    case getType(loadTodosAsync.request):
      return { ...state, todoData: loadingFlag.requested };
    case getType(loadTodosAsync.success):
      return { ...state, todoData: loadingFlag.succeded };
    case getType(loadTodosAsync.failure):
      return { ...state, todoData: loadingFlag.failed };
    case getType(postActions.loadPostsAsync.request):
      return { ...state, postData: loadingFlag.requested };
    case getType(postActions.loadPostsAsync.success):
      return { ...state, postData: loadingFlag.succeded };
    case getType(postActions.loadPostsAsync.failure):
      return { ...state, postData: loadingFlag.failed };
    case getType(postActions.addPostAsync.request):
      return { ...state, postAdd: loadingFlag.requested };
    case getType(postActions.addPostAsync.success):
      return { ...state, postAdd: loadingFlag.succeded };
    case getType(postActions.addPostAsync.failure):
      return { ...state, postAdd: loadingFlag.failed };
    default:
      return state;
  }
}
