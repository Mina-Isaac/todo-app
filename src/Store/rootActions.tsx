import * as posts from './Posts/posts.actions'
import * as todos from './Todos/todos.actions'

export default {
    posts,
    todos
}