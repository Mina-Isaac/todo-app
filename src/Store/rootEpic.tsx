import { combineEpics } from 'redux-observable';
import * as postsEpics from './Posts/posts.epic';
import * as todosEpics from './Todos/todos.epic'

export default combineEpics(...Object.values(postsEpics), ...Object.values(todosEpics))

