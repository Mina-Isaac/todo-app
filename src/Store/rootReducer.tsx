import { combineReducers } from "redux";
import postsReducer from "./Posts/posts.reducer";
import todosReducer from './Todos/todos.reducer';
import paginationReducer from './Pagination/pagination.reducer'
import loadingStateReducer from './loading.reducer'

const rootReducer = combineReducers({
    posts: postsReducer,
    todos: todosReducer,
    loadingState: loadingStateReducer,
    pagination: paginationReducer,
  });

  export default rootReducer