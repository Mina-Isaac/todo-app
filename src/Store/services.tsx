import { TODOS_API_URL, POSTS_API_URL } from "../constatnts";
import axios from "axios";
import { Post, Todo } from "../constatnts";

let postsCache: Post[] | null = null,
  todosCache: Todo[] | null = null;

const appServices = {
  getAllPosts(): Promise<Post[]> {
    if (postsCache) return new Promise((resolve) => resolve(postsCache!));
    return axios.get(POSTS_API_URL).then((response) => {
      postsCache = response.data;
      return response.data;
    });
  },

  addPost(post: Post): Promise<Post> {
    return axios.post(POSTS_API_URL, post).then((response) => response.data);
  },

  getAllTodos(): Promise<Todo[]> {
    if (todosCache) return new Promise((resolve) => resolve(todosCache!));
    return axios.get(TODOS_API_URL).then((response) => {
      todosCache = response.data;
      return response.data;
    });
  },

  editTodo(item: Todo): Promise<Todo> {
    const path = `${TODOS_API_URL}/${item.id}`;
    return axios.patch(path, item).then((response) => response.data);
  },
};

export type Services = typeof appServices;

export default appServices;
