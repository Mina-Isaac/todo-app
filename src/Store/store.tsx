import { applyMiddleware, compose, createStore, Middleware } from "redux";
import { createEpicMiddleware } from "redux-observable";
import { StateType, ActionType } from "typesafe-actions";
import appServices from "./services";
import rootReducer from "./rootReducer";
import rootAction from "./rootActions";
import rootEpic from "./rootEpic";
import {Services} from './services'

export type Action = ActionType<typeof rootAction>;
export type AppState = StateType<typeof rootReducer>;
const epicMiddleware = createEpicMiddleware<Action, Action, AppState, Services>({
  dependencies: appServices
});
const middleware: Middleware = epicMiddleware;

const enhancer = compose(applyMiddleware(middleware));
const store = createStore(rootReducer, enhancer);
epicMiddleware.run(rootEpic);

export default store;
