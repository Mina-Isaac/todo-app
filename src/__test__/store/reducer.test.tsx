import reducer from "../../Store/rootReducer";
import {AppState} from '../../Store/store'
import * as todoActions from "../../Store/Todos/todos.actions";
import * as postActions from '../../Store/Posts/posts.actions';
import sampleTodos from '../fixtures/sampleTodo'
import samplePosts from '../fixtures/samplePosts'
import {loadingFlag} from '../../Store/loading.reducer'



const getInitialState = (initial?: Partial<AppState>) =>
  reducer(initial as AppState, {} as any);

describe("Reducer test", () => {
  it("initial state should match a snapshot", () => {
    const initialState = getInitialState();
    expect(initialState).toMatchSnapshot();
  });

  it("should put retrieved todos data in the store and set loading flag to failed", () => {
    const initialState = getInitialState();
    const state = reducer(
      initialState,
      todoActions.loadTodosAsync.success(sampleTodos)
    );
    expect(state.todos).toStrictEqual(sampleTodos);
    expect(state.loadingState.todoData).toEqual(loadingFlag.succeded);
  });

  it("should put retrieved posts data in the store and set loading flag to failed", () => {
    const initialState = getInitialState();
    const state = reducer(
      initialState,
      postActions.loadPostsAsync.success(samplePosts)
    );
    expect(state.posts).toStrictEqual(samplePosts);
    expect(state.loadingState.postData).toEqual(loadingFlag.succeded);
  });
  it("should set data to [] and set loading flag to false when an error occures", () => {
    const initialState = getInitialState();
    initialState.loadingState.postData = loadingFlag.requested;
    initialState.loadingState.todoData = loadingFlag.requested;
    const stateWithPostData = reducer(
      initialState,
      postActions.loadPostsAsync.failure(new Error("some error occurred"))
    );
    const stateWithTodoData = reducer(
      initialState,
      todoActions.loadTodosAsync.failure(new Error("some error occurred"))
    );
    expect(stateWithPostData.posts).toHaveLength(0);
    expect(stateWithPostData.loadingState.postData).toEqual(loadingFlag.failed);
    expect(stateWithTodoData.todos).toHaveLength(0);
    expect(stateWithTodoData.loadingState.todoData).toEqual(loadingFlag.failed);
  });
});
