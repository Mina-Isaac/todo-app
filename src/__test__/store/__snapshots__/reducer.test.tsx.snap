// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer test initial state should match a snapshot 1`] = `
Object {
  "loadingState": Object {
    "postAdd": "unchanged",
    "postData": "unchanged",
    "todoData": "unchanged",
  },
  "pagination": Object {
    "postOffset": 0,
    "todoOffset": 0,
  },
  "posts": Array [],
  "todos": Array [],
}
`;
